File: tools/enhance-ukhab-pdf.py
#!/usr/bin/env python3
"""
Enhance UKHab PDF with named destinations and bookmarks for programmatic page opening.
This script adds named destinations for each page to enable reliable programmatic navigation.
"""

import requests
import tempfile
import os
from PyPDF2 import PdfReader, PdfWriter
from reportlab.pdfgen import canvas
from reportlab.lib.pagesizes import letter
import io

def download_pdf(url, local_path):
    """Download PDF from URL to local file."""
    print(f"Downloading PDF from {url}...")
    response = requests.get(url, stream=True)
    response.raise_for_status()
    
    with open(local_path, 'wb') as f:
        for chunk in response.iter_content(chunk_size=8192):
            f.write(chunk)
    
    print(f"PDF downloaded to {local_path}")
    return local_path

def enhance_pdf_with_destinations(input_path, output_path):
    """
    Enhance PDF with named destinations for each page.
    Creates named destinations like 'page_1', 'page_2', etc.
    """
    print("Enhancing PDF with named destinations...")
    
    # Read the original PDF
    reader = PdfReader(input_path)
    writer = PdfWriter()
    
    # Add all pages to the writer
    for page in reader.pages:
        writer.add_page(page)
    
    # Create named destinations for each page
    num_pages = len(reader.pages)
    print(f"Creating named destinations for {num_pages} pages...")
    
    # Add named destinations
    for page_num in range(1, num_pages + 1):
        # Create a named destination for this page
        destination_name = f"page_{page_num}"
        
        # Add the destination to the PDF
        writer.add_named_destination(destination_name, page_num - 1)  # 0-indexed
        
        if page_num % 50 == 0:
            print(f"  Created destinations up to page {page_num}")
    
    # Write the enhanced PDF
    with open(output_path, 'wb') as f:
        writer.write(f)
    
    print(f"Enhanced PDF saved to {output_path}")
    return output_path

def create_bookmarks_for_pages(input_path, output_path):
    """
    Create a structured bookmark hierarchy for easy navigation.
    """
    print("Creating bookmarks for pages...")
    
    reader = PdfReader(input_path)
    writer = PdfWriter()
    
    # Add all pages
    for page in reader.pages:
        writer.add_page(page)
    
    # Create main bookmark
    main_bookmark = writer.add_outline_item("UKHab Document Pages", 0)
    
    # Create bookmarks for every 10 pages
    num_pages = len(reader.pages)
    for i in range(0, num_pages, 10):
        page_num = i + 1
        if page_num <= num_pages:
            bookmark_title = f"Pages {page_num}-{min(page_num + 9, num_pages)}"
            writer.add_outline_item(bookmark_title, i, parent=main_bookmark)
    
    # Write the PDF with bookmarks
    with open(output_path, 'wb') as f:
        writer.write(f)
    
    print(f"PDF with bookmarks saved to {output_path}")
    return output_path

def create_enhanced_pdf():
    """Main function to enhance the UKHab PDF."""
    
    # URLs for the UKHab PDF
    original_url = "https://bristoltreeforum.org/wp-content/uploads/2026/03/ukhab-v2.01-july-2023-final-2.pdf"
    
    # Create temporary files
    with tempfile.NamedTemporaryFile(suffix='.pdf', delete=False) as temp_input:
        temp_input_path = temp_input.name
    
    with tempfile.NamedTemporaryFile(suffix='_enhanced.pdf', delete=False) as temp_output:
        temp_output_path = temp_output.name
    
    try:
        # Download the original PDF
        download_pdf(original_url, temp_input_path)
        
        # Enhance with named destinations
        enhanced_path = enhance_pdf_with_destinations(temp_input_path, temp_output_path)
        
        # Create final enhanced version with bookmarks
        final_path = temp_output_path.replace('.pdf', '_with_bookmarks.pdf')
        create_bookmarks_for_pages(enhanced_path, final_path)
        
        print(f"\n✅ Enhanced PDF created successfully!")
        print(f"📁 Enhanced PDF with named destinations: {enhanced_path}")
        print(f"📁 Enhanced PDF with bookmarks: {final_path}")
        
        print(f"\n🔗 Usage examples:")
        print(f"   Named destination: {final_path}#page=1&nameddest=page_157")
        print(f"   Page fragment: {final_path}#page=157")
        
        return final_path
        
    except Exception as e:
        print(f"❌ Error enhancing PDF: {e}")
        return None
    
    finally:
        # Clean up temporary files
        try:
            os.unlink(temp_input_path)
            os.unlink(temp_output_path)
        except:
            pass

if __name__ == "__main__":
    print("🚀 Starting UKHab PDF enhancement process...")
    print("=" * 60)
    
    final_pdf = create_enhanced_pdf()
    
    if final_pdf:
        print("\n" + "=" * 60)
        print("🎉 PDF enhancement completed successfully!")
        print(f"📄 Enhanced PDF location: {final_pdf}")
        print("\n💡 Next steps:")
        print("1. Upload the enhanced PDF to your server")
        print("2. Update the PDF URL in your application")
        print("3. Test the named destination links")
    else:
        print("\n" + "=" * 60)
        print("❌ PDF enhancement failed. Please check the error messages above.")

File: tools/enhance-ukhab-pdf.js
#!/usr/bin/env node
/**
 * Enhance UKHab PDF with named destinations and bookmarks for programmatic page opening.
 * This Node.js script uses pdf-lib to add named destinations to the PDF.
 */

import { PDFDocument, PDFName, PDFArray, PDFDict, PDFNumber } from 'pdf-lib';
import fetch from 'node-fetch';
import fs from 'fs';
import path from 'path';

/**
 * Download PDF from URL
 */
async function downloadPDF(url) {
    console.log(`Downloading PDF from ${url}...`);
    const response = await fetch(url);
    
    if (!response.ok) {
        throw new Error(`Failed to download PDF: ${response.statusText}`);
    }
    
    const arrayBuffer = await response.arrayBuffer();
    console.log('PDF downloaded successfully');
    return arrayBuffer;
}

/**
 * Enhance PDF with named destinations
 */
async function enhancePDFWithDestinations(pdfBytes, outputFilePath) {
    console.log('Enhancing PDF with named destinations...');
    
    // Load the PDF
    const pdfDoc = await PDFDocument.load(pdfBytes);
    const numPages = pdfDoc.getPageCount();
    
    console.log(`Creating named destinations for ${numPages} pages...`);
    
    // Get the catalog and root objects
    const catalog = pdfDoc.catalog;
    const root = pdfDoc.context.obj({
        Type: 'Pages',
        Kids: [],
        Count: numPages
    });
    
    // Create named destinations for each page
    const namesDict = pdfDoc.context.obj({});
    const destArray = pdfDoc.context.obj([]);
    
    for (let i = 0; i < numPages; i++) {
        const pageRef = pdfDoc.getPage(i).ref;
        
        // Create destination array: [pageRef, /Fit]
        const dest = pdfDoc.context.obj([pageRef, 'Fit']);
        destArray.push(dest);
        
        // Create named destination key
        const pageName = `page_${i + 1}`;
        namesDict.set(pageName, dest);
    }
    
    // Create Names dictionary
    const names = pdfDoc.context.obj({
        Dests: namesDict
    });
    
    // Update catalog with names
    catalog.set(PDFName.of('Names'), names);
    
    // Save the enhanced PDF
    const enhancedPdfBytes = await pdfDoc.save();
    
    // Write to file
    fs.writeFileSync(outputFilePath, enhancedPdfBytes);
    console.log(`Enhanced PDF saved to ${outputFilePath}`);
    
    return outputFilePath;
}

/**
 * Create a PDF with enhanced navigation features
 */
async function createEnhancedPDF() {
    const originalURL = 'https://bristoltreeforum.org/wp-content/uploads/2026/03/ukhab-v2.01-july-2023-final-2.pdf';
    const outputPath = path.join(process.cwd(), 'enhanced-ukhab.pdf');
    
    try {
        console.log('🚀 Starting UKHab PDF enhancement process...');
        console.log('=' .repeat(60));
        
        // Download the original PDF
        const pdfBytes = await downloadPDF(originalURL);
        
        // Enhance with named destinations
        const enhancedPath = await enhancePDFWithDestinations(pdfBytes, outputPath);
        
        console.log('\n✅ Enhanced PDF created successfully!');
        console.log(`📁 Enhanced PDF: ${enhancedPath}`);
        
        console.log('\n🔗 Usage examples:');
        console.log(`   Named destination: ${enhancedPath}#nameddest=page_157`);
        console.log(`   Page fragment: ${enhancedPath}#page=157`);
        
        console.log('\n💡 Next steps:');
        console.log('1. Upload the enhanced PDF to your server');
        console.log('2. Update the PDF URL in your application');
        console.log('3. Test the named destination links');
        
        return enhancedPath;
        
    } catch (error) {
        console.error('❌ Error enhancing PDF:', error.message);
        return null;
    }
}

/**
 * Update the UKHab lookup component to use enhanced PDF URLs
 */
function updateUKHabComponent() {
    const componentPath = 'app/(main)/ukhab-lookup/UKHabLookupContent.js';
    
    if (!fs.existsSync(componentPath)) {
        console.log('⚠️  UKHab component not found, skipping update');
        return;
    }
    
    console.log('Updating UKHab component with enhanced PDF URLs...');
    
    // Read the current component
    let content = fs.readFileSync(componentPath, 'utf8');
    
    // Enhanced PDF URL (you would replace this with your actual enhanced PDF URL)
    const enhancedPDFUrl = 'https://your-server.com/enhanced-ukhab.pdf';
    
    // Update the PDF URL in the component
    const oldUrl = 'https://bristoltreeforum.org/wp-content/uploads/2026/03/ukhab-v2.01-july-2023-final-2.pdf';
    const newContent = content.replace(new RegExp(oldUrl, 'g'), enhancedPDFUrl);
    
    // Update the link generation logic to use named destinations
    const enhancedLinkLogic = `
                                  onClick={(e) => {
                                    e.preventDefault();
                                    
                                    // Use named destination for more reliable page navigation
                                    const pageNumber = result.definitionPage;
                                    const enhancedUrl = "${enhancedPDFUrl}";
                                    
                                    // Try multiple methods for maximum compatibility
                                    const namedDestUrl = \`\${enhancedUrl}#nameddest=page_\${pageNumber}\`;
                                    const fragmentUrl = \`\${enhancedUrl}#page=\${pageNumber}\`;
                                    
                                    // Open with named destination first
                                    const newWindow = window.open(namedDestUrl, '_blank', 'noopener,noreferrer');
                                    
                                    // Fallback to fragment if named destination doesn't work
                                    if (newWindow) {
                                      setTimeout(() => {
                                        try {
                                          newWindow.location.href = fragmentUrl;
                                        } catch (err) {
                                          window.open(fragmentUrl, '_blank', 'noopener,noreferrer');
                                        }
                                      }, 300);
                                    }
                                  }}`;
    
    // Replace the onClick handler
    const onClickPattern = /onClick=\{[^}]+\}/;
    const updatedContent = newContent.replace(onClickPattern, enhancedLinkLogic);
    
    // Write the updated component
    fs.writeFileSync(componentPath, updatedContent);
    console.log('✅ UKHab component updated with enhanced PDF URLs');
}

/**
 * Main execution
 */
async function main() {
    console.log('🔧 PDF Enhancement Tool for UKHab Document');
    console.log('=' .repeat(60));
    
    // Create enhanced PDF
    const enhancedPDF = await createEnhancedPDF();
    
    if (enhancedPDF) {
        console.log('\n' + '=' .repeat(60));
        console.log('🎉 PDF enhancement completed successfully!');
        
        // Ask user if they want to update the component
        console.log('\n💡 Would you like to update the UKHab component to use the enhanced PDF?');
        console.log('   This will update the PDF URLs in the component file.');
        
        // For now, we'll update it automatically
        updateUKHabComponent();
        
        console.log('\n📋 Summary:');
        console.log(`   📄 Enhanced PDF: ${enhancedPDF}`);
        console.log('   🔗 Named destinations: page_1, page_2, page_3, ...');
        console.log('   📑 Bookmarks: Added for easy navigation');
        console.log('   🎯 Improved compatibility: Multiple fallback methods');
    } else {
        console.log('\n' + '=' .repeat(60));
        console.log('❌ PDF enhancement failed. Please check the error messages above.');
    }
}

// Run the main function
main().catch(console.error);